### ISSUE TITLE
support offline enabled EOL Taxon matching

### ISSUE THREAD

**Issue by jhpoelen** [globalbioticinteractions/nomer#11]
Now that EOL publishes a taxon hierarchy (see https://github.com/jhpoelen/eol-globi-data/issues/384), an offline enabled EOL taxon page lookup can be created. In contrast with a taxon-by-taxon web api search, an archive based integration is expected to be stable, offline-enabled and work well with large datasets.

**Comment by jhpoelen** [comment]
@jhammock  - please holler in this thread once the EOL dynamic hierarchy is available via https://opendata.eol.org .

**Comment by jhammock** [comment]
https://opendata.eol.org/dataset/identifier-map

Please bear with us while we figure out how to handle versioning...

**Comment by jhpoelen** [comment]
Great! This covers the id mapping part, or the connections/edges between taxon concepts across name authorities (e.g., the taxonMap of the https://doi.org/10.5281/zenodo.755513). There are two eol ids (e.g., page and node). Currently, GloBI is using the page ids. Is this something you'd recommend going forward?

**Comment by jhammock** [comment]
Yes, the node_id is apparently an internal thing that can be useful with some of our APIs, but is not really meant to be outward facing. The page_id is what you want. 

Note that the resource_pks from the EOL Dynamic Hierarchy are not yet stable; those negative values were a byproduct that was unpopular. They'll change in the next version and be stable thereafter. But I don't think you had an immediate need for those...

**Comment by jhpoelen** [comment]
Just curious - to you keep a list around of which source names (e.g., scientific names found in trait dataset) were mapped to what ids?

**Comment by jhammock** [comment]
Hmm... we haven't made a file, though the API has that. Is there a use case?

**Comment by jhpoelen** [comment]
In my mind, without such a map, the name resolution in EOL is a black box and is hard to version / reproduce / control. Also, having such a map makes it easier to decouple the name resolution process and make for a more modular integration process. In addition, others can re-use the map as is to avoid having to rebuild some complicate name resolution process. 

So, use cases involve software architecture, maintainability, change control and quality control.

**Comment by jhammock** [comment]
Why namestrings rather than IDs? Or do you mean in particular for resources that don't maintain IDs internally? I think I may be confounding your use cases.

**Comment by jhpoelen** [comment]
Just to document the claims that EOL makes about incoming data sources with or without taxonomic identifiers and how this maps into the EOL universe. 

For instance, 
trait dataset A may say "Homo sapiens" . On ingestion, EOL would make a claim that "Homo sapiens" (no id) is equivalent to "Homo sapiens" (EOL:1234). 

Also, 
name dataset B may say "Homo sapiens" (SOME:2345), On ingestion, EOL would make a claim that "Homo sapiens" (SOME:2345) is equivalent to "Homo sapiens" (EOL:1234).

I think it is relevant to explicitly publish these claims, because taxonomic name mappings are known to be error prone, subjective and hard to reproduce.

Is this what you had in mind?

**Comment by jhammock** [comment]
I follow you now, I think. Yes, we're hoping to fully split out the names matching part of the harvest process, (we just got the functionality of bypassing it and doing the rest of a harvest, with pre-assigned EOL IDs). Then we should be able to make it fairly easy, for instance, for anyone to run a names file against the current code, either a names file from one of our [content partner resources](https://opendata.eol.org/organization/eol-content-partners) or a novel one, without our intervention. Or whatever kind of interface users are looking for for examining names matches in bulk. Nothing gets prioritized yet without an external use case :)

Meanwhile, names mappings per resource per taxon are individually published, in the API and [the interface](https://eol.org/pages/311544/names)

**Comment by jhpoelen** [comment]
Please let me know If you need me to clarify the external use cases I presented in this thread for publishing name mapping archives suitable for quality / integration processes at scale. 

So, if I understand it correctly, EOL currently provides a id-to-id mapping archive from name providers to EOL. This mapping archive contains unstable name provider ids. Also, the taxonomic information associated with the EOL taxon names will be available at some point when the EOL Dynamic Hierarchy archive is available for download.

**Comment by jhpoelen** [comment]
@jhammock I am looking at some other offline taxonomic matching methods and re-visited the EOL taxon resource. Is http://content.eol.org/resources/817 with http://content.eol.org/data/dvdtg/taxon.tab the resource for GloBI to look at?

**Comment by jhpoelen** [comment]
Via @jhammock I've learned that https://opendata.eol.org/dataset/tram-807-808-809-810-dh-v1-1/resource/00adb47b-57ed-4f6b-8f66-83bfdb5120e8 is the most recent copy of EOL's dynamic hierarchy.

**Comment by jhpoelen** [comment]
See also https://github.com/eol/harvester and https://github.com/EOL/harvester/blob/master/app/models/names_matcher.rb for method on how EOL links to their dynamic hierarchy to names coming from external data sources.

**Comment by jhpoelen** [comment]
A first pass at implementing enabling offline matching against EOL Dynamic Hierarchy uncovered a 403 (forbidden) error code when Nomer attempted to download https://editors.eol.org/other_files/DWH/TRAM-809/DH_v1_1.tar.gz (see error log below). Confusingly, the same url could be accessed using command-line tools like ```wget``` and ```curl```. To work around this problem, a repackaged Zenodo publication for the EOL Dynamic Hierarchy was created:

 Encyclopedia of Life. (2020). Repackaged Encyclopedia of Life (EOL) Dynamic Hierarchy (Version 1.1) [Data set]. Zenodo. http://doi.org/10.5281/zenodo.3834881

EOL team is being notified of this.

```
Caused by: org.apache.commons.vfs2.FileSystemException: GET method failed for "https://editors.eol.org/other_files/DWH/TRAM-809/DH_v1_1.tar.gz" with HTTP status 403.
	at org.apache.commons.vfs2.provider.http.HttpFileObject.doGetInputStream(HttpFileObject.java:128)
	at org.apache.commons.vfs2.provider.AbstractFileObject.getInputStream(AbstractFileObject.java:1119)
	at org.apache.commons.vfs2.provider.DefaultFileContent.getInputStream(DefaultFileContent.java:349)
	at org.apache.commons.vfs2.provider.gzip.GzipFileObject.doGetInputStream(GzipFileObject.java:50)
	at org.apache.commons.vfs2.provider.AbstractFileObject.getInputStream(AbstractFileObject.java:1119)
	at org.apache.commons.vfs2.provider.DefaultFileContent.getInputStream(DefaultFileContent.java:349)
	at org.apache.commons.vfs2.provider.DefaultFileContent.write(DefaultFileContent.java:734)
	at org.apache.commons.vfs2.provider.DefaultFileContent.write(DefaultFileContent.java:720)
	at org.apache.commons.vfs2.provider.DefaultFileContent.write(DefaultFileContent.java:691)
	at org.apache.commons.vfs2.provider.DefaultFileContent.write(DefaultFileContent.java:707)
	at org.apache.commons.vfs2.FileUtil.copyContent(FileUtil.java:78)
	at org.apache.commons.vfs2.provider.AbstractFileObject.copyFrom(AbstractFileObject.java:289)
```

**Comment by jhpoelen** [comment]
First pass at indexing EOL taxonomy (initial download/index takes 45 minutes), subsequent lookups are << 1 second:

```
$ echo "EOL:327955" | nomer append eol-taxon-id
using matcher [eol-taxon-id]
EOL taxonomy importing...
caching [gz:https://zenodo.org/record/3834881/files/taxon.tab.gz!/taxon.tab] at [/media/jorrit/data/nomer/6b45b2ea7e9afaa348e859f6ad70362bc95d9eecd66b4a72e7b6885e146526e4.gz]...
caching [gz:https://zenodo.org/record/3834881/files/taxon.tab.gz!/taxon.tab] at [/media/jorrit/data/nomer/6b45b2ea7e9afaa348e859f6ad70362bc95d9eecd66b4a72e7b6885e146526e4.gz] done.
using cached [gz:https://zenodo.org/record/3834881/files/taxon.tab.gz!/taxon.tab] at [/media/jorrit/data/nomer/6b45b2ea7e9afaa348e859f6ad70362bc95d9eecd66b4a72e7b6885e146526e4.gz]
cache with [4011067] items built in [2723.7] s or [1472.6] items/s.
EOL taxonomy imported.
EOL:327955	SAME_AS	EOL:327955	Homo sapiens	species		Life | Cellular Organisms | Eukaryota | Opisthokonta | Metazoa | Bilateria | Deuterostomia | Chordata | Vertebrata | Gnathostomata | Osteichthyes | Sarcopterygii | Tetrapoda | Amniota | Synapsida | Therapsida | Cynodontia | Mammalia | Theria | Eutheria | Placentalia | Boreoeutheria | Euarchontoglires | Euarchonta | Primates | Haplorrhini | Anthropoidea | Catarrhini | Hominoidea | Hominidae | Homininae | Homo | Homo sapiens	EOL:2913056 | EOL:6061725 | EOL:2908256 | EOL:2910700 | EOL:1 | EOL:3014411 | EOL:8814528 | EOL:694 | EOL:2774383 | EOL:46557929 | EOL:2775704 | EOL:8909 | EOL:46557930 | EOL:3014446 | EOL:3014447 | EOL:4522847 | EOL:4528578 | EOL:1642 | EOL:47048248 | EOL:47048864 | EOL:42399417 | EOL:32005829 | EOL:40107874 | EOL:42410181 | EOL:1645 | EOL:10487985 | EOL:42410186 | EOL:8824215 | EOL:4529848 | EOL:47049573 | EOL: | EOL:42268 | EOL:327955	clade | clade | domain | clade | kingdom | clade | clade | phylum | subphylum | clade | clade | class | clade | clade | clade | order | suborder | class | clade | clade | clade | clade | clade | clade | order | suborder | clade | clade | superfamily | family | subfamily | genus | species	http://eol.org/pages/327955	
```

Preliminary results suggest that Nomer is now able to match at a rate of about about 3k ids per second are matched (e.g., 111k EOL ids in 37s)

```
$ time zcat taxonCache.tsv.gz | grep EOL | cut -f1 | nomer append eol-taxon-id | pv -l > /dev/null
using matcher [eol-taxon-id]
EOL taxonomy already indexed at [/media/jorrit/data/nomer/eol/eol], no need to import.
 111k 0:00:37 [2.95k/s] [                                <=>                   ]

real	0m37.702s
user	0m58.011s
sys	0m5.318s
```

**Comment by jhpoelen** [comment]
first pass at offline-enable EOL matching included in Nomer v0.1.11 .
